contracts/src/v0.8/upkeeps/GovernanceUpkeep.sol

```diff
@@ -0,0 +1,137 @@
+import "../interfaces/KeeperCompatibleInterface.sol";
+import "../vendor/GovernorAlpha.sol";
+
+/// @notice Possible actions that can be taken in the performUpkeep function.
+/// QUEUE => calls 'queue(id)' on the governance contract
+/// EXECUTE => calls 'execute(id)' on the governance contract
+/// CANCEL => calls 'cancel(id)' on the governance contract
+/// UPDATE_INDEX => updates the starting proposal index within the
+///                 upkeep contract to reduce the amount of proposals
+///                 the need to be checked
+enum Action {
+  QUEUE,
+  EXECUTE,
+  CANCEL,
+  UPDATE_INDEX
+}
+
+/// @title Chainlink Keepers Compatible GovernorAlpha Automator
+contract GovernanceAutomator is KeeperCompatibleInterface {
+  GovernorAlpha public immutable s_governanceContract;
+  GovernorAlphaToken public immutable s_governanceTokenContract;
+  uint256 public s_proposalStartingIndex;
+  Action public action;
+
+  constructor(
+    GovernorAlpha _governanceContract,
+    uint256 _proposalStartingIndex,
+    address _tokenContract
+  ) {
+    s_governanceContract = _governanceContract;
+    s_proposalStartingIndex = _proposalStartingIndex;
+    s_governanceTokenContract = GovernorAlphaToken(_tokenContract);
+  }
+
+  ///@notice Simulated at each block by the Chainlink Keepers network. Checks if there are any actions (queue() or execute()) required on a governance contract. Also tracks a 'starting index'.
+  ///@return upkeepNeeded return true if performUpkeep should be called
+  ///@return performData bytes encoded: (governance action required, index of proposal)
+  function checkUpkeep(
+    bytes calldata /* checkData */
+  ) external view override returns (bool upkeepNeeded, bytes memory performData) {
+    //Get number of proposal
+    uint256 proposalCount = s_governanceContract.proposalCount();
+
+    //Find starting index
+    uint256 newStartingIndex = findStartingIndex();
+
+    //If new starting index found, update in performUpkeep
+    if (newStartingIndex > s_proposalStartingIndex) {
+      performData = abi.encode(Action.UPDATE_INDEX, newStartingIndex);
+      return (true, performData);
+    }
+
+    //Go through each proposal and check the current state
+    for (uint256 i = newStartingIndex; i <= proposalCount; i++) {
+      GovernorAlpha.ProposalState state = s_governanceContract.state(i);
+      (, address proposer, , , , , , , ) = s_governanceContract.proposals(i);
+
+      if (state == GovernorAlpha.ProposalState.Succeeded) {
+        //If the state is 'Succeeded' then call 'queue' with the Proposal ID
+        performData = abi.encode(Action.QUEUE, i);
+        return (true, performData);
+      } else if (state == GovernorAlpha.ProposalState.Queued) {
+        //If the state is 'Queued' then call 'execute' with the Proposal ID
+        performData = abi.encode(Action.EXECUTE, i);
+        return (true, performData);
+      } else if (
+        s_governanceTokenContract.getPriorVotes(proposer, sub256(block.number, 1)) <
```

Comment:
should we consider the ProposalState here? ex we should only try to cancel proposals who's proposal state is `Active` or `Defeated` or something?